File: lib/contacts.c
//
//  General-purpose Photovoltaic Device Model gpvdm.com- a drift diffusion
//  base/Shockley-Read-Hall model for 1st, 2nd and 3rd generation solarcells.
// 
//  Copyright (C) 2012 Roderick C. I. MacKenzie r.c.i.mackenzie at googlemail.com
//
//	https://www.gpvdm.com
//	Room B86 Coates, University Park, Nottingham, NG7 2RD, UK
//
//
// This program is free software; you can redistribute it and/or modify it
// under the terms and conditions of the GNU General Public License,
// version 2, as published by the Free Software Foundation.
//
// This program is distributed in the hope it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
// more details.


#include <string.h>
#include "epitaxy.h"
#include "inp.h"
#include "util.h"
#include "const.h"
#include <cal_path.h>
#include "contacts.h"

void contacts_time_step(struct simulation *sim,struct device *in)
{
	int i;

	for (i=0;i<in->ncontacts;i++)
	{
		in->contacts[i].voltage_last=in->contacts[i].voltage;
	}
}

void contacts_load(struct simulation *sim,struct device *in)
{
	int i;
	struct inp_file inp;
	in->ncontacts=0;

	inp_init(sim,&inp);
	if (inp_load(sim, &inp , "contacts.inp")!=0)
	{
		ewe(sim,"Can't open the file contacts\n");
	}

	inp_check(sim,&inp,1.1);
	inp_reset_read(sim,&inp);
	inp_get_string(sim,&inp);
	sscanf(inp_get_string(sim,&inp),"%d",&(in->ncontacts));

	if (in->ncontacts>10)
	{
		ewe(sim,"Too many contacts\n");
	}

	if (in->ncontacts<1)
	{
		ewe(sim,"No contacts\n");
	}

	gdouble pos=0.0;
	int active=FALSE;
	for (i=0;i<in->ncontacts;i++)
	{
		
		inp_get_string(sim,&inp);	//active contact
		strcpy(in->contacts[i].name,inp_get_string(sim,&inp));

		inp_get_string(sim,&inp);	//position
		in->contacts[i].position=english_to_bin(sim, inp_get_string(sim,&inp));
	
		inp_get_string(sim,&inp);	//active contact
		in->contacts[i].active=english_to_bin(sim, inp_get_string(sim,&inp));

		inp_get_string(sim,&inp);	//start
		sscanf(inp_get_string(sim,&inp),"%Le",&(in->contacts[i].start));

		inp_get_string(sim,&inp);	//width
		sscanf(inp_get_string(sim,&inp),"%Le",&(in->contacts[i].width));

		inp_get_string(sim,&inp);	//depth
		sscanf(inp_get_string(sim,&inp),"%Le",&(in->contacts[i].depth));

		inp_get_string(sim,&inp);	//voltage
		sscanf(inp_get_string(sim,&inp),"%Le",&(in->contacts[i].voltage));
		in->contacts[i].voltage_last=in->contacts[i].voltage;


		pos+=in->contacts[i].width;
	}

	char * ver = inp_get_string(sim,&inp);
	if (strcmp(ver,"#ver")!=0)
	{
			ewe(sim,"No #ver tag found in file\n");
	}

	inp_free(sim,&inp);

	contacts_update(sim,in);
}

void contacts_force_to_zero(struct simulation *sim,struct device *in)
{
int x;
int z;

for (x=0;x<in->xmeshpoints;x++)
{
	for (z=0;z<in->zmeshpoints;z++)
	{
		in->Vapplied_l[z][x]=0.0;
		in->Vapplied_r[z][x]=0.0;
	}

}

}

void contacts_dump(struct device *in)
{
int i;
	for (i=0;i<in->ncontacts;i++)
	{
		printf("%Le\n",in->contacts[i].voltage);
	}
	
}

void contacts_update(struct simulation *sim,struct device *in)
{
int i;
int x;
int z;
int n;
int found=FALSE;

gdouble value=0.0;

if (in->xmeshpoints==1)
{
	for (z=0;z<in->zmeshpoints;z++)
	{
		for (i=0;i<in->ncontacts;i++)
		{
			if (in->contacts[i].position==TOP)
			{
				in->Vapplied_r[z][0]=in->contacts[i].voltage;
			}else
			{
				in->Vapplied_l[z][0]=in->contacts[i].voltage;
			}
		}
	}

	return;
}

//Contacts on top
for (x=0;x<in->xmeshpoints;x++)
{
	found=FALSE;
	for (i=0;i<in->ncontacts;i++)
	{
		if ((in->xmesh[x]>=in->contacts[i].start)&&(in->xmesh[x]<in->contacts[i].start+in->contacts[i].width))
		{
			if (in->contacts[i].position==TOP)
			{
				value=in->contacts[i].voltage;
				n=i;
				found=TRUE;
				break;
			}
		}
	}

	if (found==FALSE)
	{
		value=0.0;
		n=-1;
	}

	for (z=0;z<in->zmeshpoints;z++)
	{
		in->Vapplied_r[z][x]=value;
		in->n_contact_r[z][x]=n;
	}
	
}

//Contacts on btm
for (x=0;x<in->xmeshpoints;x++)
{
	found=FALSE;
	for (i=0;i<in->ncontacts;i++)
	{
		if ((in->xmesh[x]>=in->contacts[i].start)&&(in->xmesh[x]<in->contacts[i].start+in->contacts[i].width))
		{
			if (in->contacts[i].position==BOTTOM)
			{
				value=in->contacts[i].voltage;
				n=i;
				found=TRUE;
				break;
			}
		}
	}

	if (found==FALSE)
	{
		value=0.0;
		n=-1;
	}

	for (z=0;z<in->zmeshpoints;z++)
	{
		in->Vapplied_l[z][x]=value;
		in->n_contact_l[z][x]=n;
	}
	
}
}

gdouble contact_get_voltage_last(struct simulation *sim,struct device *in,int contact)
{
	return in->contacts[contact].voltage_last;
}

gdouble contact_get_voltage(struct simulation *sim,struct device *in,int contact)
{
	return in->contacts[contact].voltage;
}

void contact_set_voltage(struct simulation *sim,struct device *in,int contact,gdouble voltage)
{
	in->contacts[contact].voltage=voltage;
	contacts_update(sim,in);
}

void contact_set_active_contact_voltage(struct simulation *sim,struct device *in,gdouble voltage)
{
	int i=0;

	for (i=0;i<in->ncontacts;i++)
	{
		if (in->contacts[i].active==TRUE)
		{
			in->contacts[i].voltage=voltage;
		}
	}

	contacts_update(sim,in);

}

long double contact_get_active_contact_voltage(struct simulation *sim,struct device *in)
{
	int i=0;

	if ((in->xmeshpoints==1)&&(in->zmeshpoints==1))
	{
		return in->contacts[0].voltage;

	}else
	{
		for (i=0;i<in->ncontacts;i++)
		{
			if (in->contacts[i].active==TRUE)
			{
				return in->contacts[i].voltage;
			}
		}
	}

}

void contact_set_all_voltages(struct simulation *sim,struct device *in,gdouble voltage)
{
int i;
	for (i=0;i<in->ncontacts;i++)
	{
		in->contacts[i].voltage=voltage;
	}

	contacts_update(sim,in);
}

long double contacts_get_Jleft(struct device *in)
{
int i;
int x;
int z;

long double tot=0.0;
long double count=0.0;

for (x=0;x<in->xmeshpoints;x++)
{
		for (z=0;z<in->zmeshpoints;z++)
		{
			if (in->n_contact_l[z][x]>=0)
			{
				tot+=in->Jpleft[z][x]+in->Jnleft[z][x];
				count=count+1.0;						//this will need updating for meshes which change
			}
		}
}

tot=tot/count;

return tot*Q;
}

long double contacts_get_Jright(struct device *in)
{
int i;
int x;
int z;

long double tot=0.0;
long double count=0.0;

for (x=0;x<in->xmeshpoints;x++)
{
		for (z=0;z<in->zmeshpoints;z++)
		{
			if (in->n_contact_r[z][x]>=0)
			{
				tot+=in->Jpright[z][x]+in->Jnright[z][x];
				count=count+1.0;
			}
		}
}

tot=tot/count;

return tot*Q;
}


long double contacts_get_J(struct device *in, int n)
{
int i;
int x;
int z;

long double tot=0.0;
long double count=0.0;

for (x=0;x<in->xmeshpoints;x++)
{
		for (z=0;z<in->zmeshpoints;z++)
		{
			for (i=0;i<in->ncontacts;i++)
			{
				if (in->n_contact_r[z][x]==n)
				{
					tot+=in->Jpright[z][x]+in->Jnright[z][x];
					count=count+1.0;						//this will need updating for meshes which change
				}
				
				if (in->n_contact_l[z][x]==n)
				{
					tot+=in->Jpleft[z][x]+in->Jnleft[z][x];
					count=count+1.0;						//this will need updating for meshes which change
				}
			}
		}
}

tot=tot/count;

return tot*Q;
}

void contacts_passivate(struct simulation *sim,struct device *in)
{
int i;
int x;
int y;
int z;

//passivate under each contact
for (x=0;x<in->xmeshpoints;x++)
{
	for (y=0;y<in->ymeshpoints;y++)
	{
		for (z=0;z<in->zmeshpoints;z++)
		{

			for (i=0;i<in->ncontacts;i++)
			{
				if (in->contacts[i].position==TOP)
				{
					if ((in->ylen-in->ymesh[y]<=in->contacts[i].depth)&&(in->xmesh[x]>in->contacts[i].start)&&(in->xmesh[x]<in->contacts[i].start+in->contacts[i].width))
					{
						in->mun[z][x][y]=1e-15;
						in->mup[z][x][y]=1e-15;
					}
				}

				if (in->contacts[i].position==BOTTOM)
				{
					if ((in->ymesh[y]<=in->contacts[i].depth)&&(in->xmesh[x]>in->contacts[i].start)&&(in->xmesh[x]<in->contacts[i].start+in->contacts[i].width))
					{
						in->mun[z][x][y]=1e-15;
						in->mup[z][x][y]=1e-15;
					}
				}
			}
		}
	}
}

for (x=0;x<in->xmeshpoints;x++)
{
	for (z=0;z<in->zmeshpoints;z++)
	{
		i=in->n_contact_r[z][x];
		if (i==-1)
		{
			in->mun[z][x][in->ymeshpoints-1]=1e-15;
			in->mup[z][x][in->ymeshpoints-1]=1e-15;
		}

		i=in->n_contact_l[z][x];
		if (i==-1)
		{
			in->mun[z][x][0]=1e-15;
			in->mup[z][x][0]=1e-15;
		}
	}
}

}

File: gui/contacts.py
#    General-purpose Photovoltaic Device Model - a drift diffusion base/Shockley-Read-Hall
#    model for 1st, 2nd and 3rd generation solar cells.
#    Copyright (C) 2012-2017 Roderick C. I. MacKenzie r.c.i.mackenzie at googlemail.com
#
#	https://www.gpvdm.com
#	Room B86 Coates, University Park, Nottingham, NG7 2RD, UK
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License v2.0, as published by
#    the Free Software Foundation.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License along
#    with this program; if not, write to the Free Software Foundation, Inc.,
#    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.



import os
from numpy import *
import webbrowser
from inp import inp_search_token_value
from icon_lib import icon_get

from epitaxy import epitaxy_get_layers
from epitaxy import epitaxy_get_dos_file
from epitaxy import epitaxy_get_width
from scan_item import scan_item_add

import i18n
_ = i18n.language.gettext


#contacts io
from contacts_io import segment
from contacts_io import contacts_save
from contacts_io import contacts_get_array
from contacts_io import contacts_clear
from contacts_io import contacts_print
from contacts_io import contacts_load
from contacts_io import contacts_print
from contacts_io import contacts_append

#qt
from PyQt5.QtWidgets import QMainWindow, QTextEdit, QAction, QApplication
from PyQt5.QtGui import QIcon
from PyQt5.QtCore import QSize, Qt 
from PyQt5.QtWidgets import QWidget,QSizePolicy,QHBoxLayout,QPushButton,QDialog,QFileDialog,QToolBar, QMessageBox, QVBoxLayout, QGroupBox, QTableWidget,QAbstractItemView, QTableWidgetItem, QComboBox

from PyQt5.QtCore import pyqtSignal

from gui_util import tab_get_value
from gui_util import tab_set_value
from gui_util import tab_insert_row
from gui_util import tab_remove

from util import str2bool
from error_dlg import error_dlg

from global_objects import global_object_run

from QComboBoxLang import QComboBoxLang
from QWidgetSavePos import QWidgetSavePos

class contacts_window(QWidgetSavePos):

	visible=1
	
	changed = pyqtSignal()
	
	def update_contact_db(self):
		for i in range(0,self.tab.rowCount()):
			try:
				float(tab_get_value(self.tab,i, 3))
				float(tab_get_value(self.tab,i, 4))
				float(tab_get_value(self.tab,i, 5))
				float(tab_get_value(self.tab,i, 6))
			except:
				return False

		contacts_clear()
		for i in range(0,self.tab.rowCount()):
			contacts_append(tab_get_value(self.tab,i, 0),tab_get_value(self.tab,i, 1),str2bool(tab_get_value(self.tab,i, 2)),float(tab_get_value(self.tab,i, 3)),float(tab_get_value(self.tab,i, 4)),float(tab_get_value(self.tab,i, 5)),float(tab_get_value(self.tab,i, 6)))
		return True
	
	def add_row(self,pos,name,top_btm,active,start,width,depth,voltage):

		pos= tab_insert_row(self.tab)

		self.tab.blockSignals(True)
		self.tab.setItem(pos,0,QTableWidgetItem(name))

		combobox = QComboBoxLang()
		combobox.addItemLang("top",_("top"))
		combobox.addItemLang("bottom",_("bottom"))

		self.tab.setCellWidget(pos,1, combobox)
		combobox.setValue_using_english(top_btm.lower())
		#combobox.setCurrentIndex(combobox.findText(top_btm.lower()))
		combobox.currentIndexChanged.connect(self.save)

		combobox = QComboBoxLang()
		combobox.addItemLang("true",_("true"))
		combobox.addItemLang("false",_("false"))

		self.tab.setCellWidget(pos,2, combobox)
		combobox.setValue_using_english(active.lower()) #setCurrentIndex(combobox.findText(active.lower()))
		combobox.currentIndexChanged.connect(self.save)
		
		self.tab.setItem(pos,3,QTableWidgetItem(start))
		self.tab.setItem(pos,4,QTableWidgetItem(width))
		self.tab.setItem(pos,5,QTableWidgetItem(depth))
		self.tab.setItem(pos,6,QTableWidgetItem(voltage))


		self.tab.blockSignals(False)
		
	def on_add_clicked(self, button):
		index = self.tab.selectionModel().selectedRows()

		if len(index)>0:
			pos=index[0].row()+1
		else:
			pos = self.tab.rowCount()

		self.add_row(pos,"","top","false","0.0","0.0","0.0","0.0")
 
		self.save()

	def on_remove_clicked(self, button):
		tab_remove(self.tab)
		self.save()

	def save(self):
		if self.update_contact_db()==True:
			contacts_save()
			self.changed.emit()
			global_object_run("gl_force_redraw")
		else:
			error_dlg(self,_("There are some non numeric values in the table"))


	def callback_help(self):
		webbrowser.open('http://www.gpvdm.com/man/index.html')

	def tab_changed(self, x,y):
		self.save()

	def load(self):
		self.tab.clear()
		self.tab.setHorizontalHeaderLabels([_("Name"),_("Top/Bottom"),_("Active contact"),_("Start"), _("Width"),_("Depth"),_("Voltage")])
		contacts_load()
		#contacts_print()
		contacts=contacts_get_array()
		i=0
		for c in contacts_get_array():
			self.add_row(i,str(c.name),str(c.position),str(c.active),str(c.start),str(c.width),str(c.depth),str(c.voltage))

			i=i+1


	def __init__(self):
		QWidgetSavePos.__init__(self,"contacts")
		self.setFixedSize(750, 400)

		self.setWindowIcon(icon_get("contact"))

		self.setWindowTitle(_("Edit contacts")+" (www.gpvdm.com)") 
		
		self.main_vbox = QVBoxLayout()

		toolbar=QToolBar()
		toolbar.setIconSize(QSize(48, 48))

		add = QAction(icon_get("list-add"),  _("Add contact"), self)
		add.triggered.connect(self.on_add_clicked)
		toolbar.addAction(add)

		remove = QAction(icon_get("list-remove"),  _("Remove contacts"), self)
		remove.triggered.connect(self.on_remove_clicked)
		toolbar.addAction(remove)

		spacer = QWidget()
		spacer.setSizePolicy(QSizePolicy.Expanding, QSizePolicy.Expanding)
		toolbar.addWidget(spacer)


		self.help = QAction(icon_get("help"), _("Help"), self)
		self.help.setStatusTip(_("Close"))
		self.help.triggered.connect(self.callback_help)
		toolbar.addAction(self.help)

		self.main_vbox.addWidget(toolbar)

		self.tab = QTableWidget()
		self.tab.resizeColumnsToContents()

		self.tab.verticalHeader().setVisible(False)

		self.tab.clear()
		self.tab.setColumnCount(7)
		self.tab.setSelectionBehavior(QAbstractItemView.SelectRows)

		self.load()

		self.tab.cellChanged.connect(self.tab_changed)

		self.main_vbox.addWidget(self.tab)


		self.setLayout(self.main_vbox)

